feat: add docs for 'polymorphism in C++' (#311)
Codecademy/docs: content/cpp/concepts/polymorphism/polymorphism.md

---
Title: 'Polymorphism'

> Description: 'The word Polymorphism means having many form. Polymorphism is an important concept of object-oriented programming. It simply means more than one form. That is, the same entity (function or operator) can behave differently in different scenarios.'

Reviewer: ```suggestion
Description: 'Polymorphism is an important concept of object-oriented programming. It simply means more than one form. That is, the same entity (function or operator) can behave differently in different scenarios.'
```

Author: made the requested changes successfully.

Subjects:
  - 'Computer Science'
  - 'Game Development'
Tags:
  - 'Function'
  - 'Polymorphism'
CatalogContent:
  - 'learn-c-plus-plus'
  - 'paths/computer-science'
---

Polymorphism is an important concept of object-oriented programming. It simply means more than one form. That is, the same entity (function or operator) can operate differently under different situations.

> In C++ we have two types of polymorphism:

Reviewer: ```suggestion
In C++, we have two types of polymorphism:
```

Author: made the requested changes successfully.

> - Compile time Polymorphism – This is also called as static (or early) binding.

Reviewer: ```suggestion
- Compile-time polymorphism: This is also called static (or early) binding.
```

> - Runtime Polymorphism – This is also called as dynamic (or late) binding.

Reviewer: ```suggestion
- Runtime polymorphism: This is also called dynamic (or late) binding.
```

Author: made the requested changes successfully.

> ## Compile time Polymorphism

Author: made the requested changes successfully.

> Compile time polymorphism is perfect for **function overloading** as well as **operator overloading**.

Reviewer: ```suggestion
Compile-time polymorphism is perfect for function overloading as well as operator overloading.
```

Author: made the requested changes successfully.

> ### Compile time Polymorphism Example

Reviewer: ```suggestion
```

Author: made the requested changes successfully.

> In below example, we have two functions with same name but different number of arguments. Based on the number of parameters we pass during function invocation (function calling) determines which function is to be called, this is why it is considered as an example of **polymorphism** because in different conditions the output is different. Since, the call is determined during **compile time** thats why it is called **compile time polymorphism**.

Reviewer: ```suggestion
In the below example, there are two functions with the same name but a different number of arguments. 

Based on the number of parameters we pass during function invocation (function calling) determines which function is to be called, this is why it is considered as an example of polymorphism because in different conditions the output is different. Since, the call is determined during compile time that's why it is called compile-time polymorphism.
```

Author: made the requested changes successfully.

```cpp
#include <iostream>

> using namespace std;

Reviewer: ```suggestion
```

class Add {
  public:

>     // function sum with 2 parametes

Reviewer: ```suggestion
    // Function sum() with 2 parameters
```

Author: made the requested changes successfully.

    int sum(int num1, int num2){
      return num1 + num2;
    }

>     // function sum with 3 parametes

Reviewer: ```suggestion
    // Function sum() with 3 parameters
```

Author: made the requested changes successfully.

    int sum(int num1, int num2, int num3){
      return num1 + num2 + num3;
    }
};

int main() {

>   // create object of Add class

Reviewer: ```suggestion
  // Create object of Add class
```

Author: made the requested changes successfully.

  Add obj;

>   //This will call the first variant of sum function (sum function with 2 parameters)

Reviewer: ```suggestion
  // This will call the first variant of sum() function (the one with 2 parameters)
```

Author: made the requested changes successfully.

>   cout<<"Output: "<<obj.sum(10, 20)<<endl;

Reviewer: - Write out `std::cout`
- Add spaces.
- Add "\n"

```suggestion
  std::cout << "Output: " << obj.sum(10, 20) << "\n";
```

Author: made the requested changes successfully.

>   //This will call the second variant of sum second function (sum function with 3 parameters)

Reviewer: ```suggestion
  // This will call the second variant of sum() function (the one with 3 parameters)
```

Author: made the requested changes successfully.

>   cout<<"Output: "<<obj.sum(11, 22, 33);

Reviewer: ```suggestion
  std::cout << "Output: " << obj.sum(11, 22, 33);
```

  return 0;
}
```

Above code snippet will generated below output.

```cpp
Output: 30 // result of first function invocation
Output: 66 // result of second function invocation
```

## Runtime Polymorphism

> **Function overriding** is an example of Runtime polymorphism.

Reviewer: Take out all the bolding of keywords.

Author: made the requested changes successfully.

### Function overriding

It is a concept when child class declares a method, which is already present in the parent class then this is called function overriding, here child class overrides the parent class.

In case of function overriding we have two definitions of the same function, one is parent class and one in child class. The call to the function is determined at **runtime** to decide which definition of the function is to be called, thats the reason it is called **runtime polymorphism**.

### Runtime Polymorphism Example

```cpp

> #include <iostream>

Reviewer: Take out `using namespace std;`

Use `std::cout` and "\n" instead of `endl` to match Codecademy style guide.

Author: made the requested changes successfully.
